lib/std/crypto/pem.zig

```diff
@@ -0,0 +1,540 @@
+    var rest = data;
+    
+    while (true) {
+        if (hasPrefix(rest, pem_start[1..])) {
+            rest = rest[pem_start.len-1..];
+        } else {
+            const cut_data = cut(rest, pem_start);
+            if (cut_data.found) {
+                rest = cut_data.after;
+            } else {
+                return Error.NotPemData;
+            }
+        }
+        
+        const line_data = getLine(rest);
+        if (!hasSuffix(line_data.line, pem_end_of_line)) {
+            continue;
+        }
+        
+        const type_line = line_data.line[0 .. line_data.line.len-pem_end_of_line.len];
+
+        rest = line_data.rest;
+        
+        var p = Block.init(allocator);
+        p.type = type_line;
+        
+        while (true) {
+            if (rest.len == 0) {
+                return Error.PemDataEmpty;
+            }
+            
+            const line_data2 = getLine(rest);
+            
+            const cut_data = cut(line_data2.line, colon);
+            if (!cut_data.found) {
+                break;
+            }
+            
+            const key = trimSpace(cut_data.before);
+            const val = trimSpace(cut_data.after);
+           
+            try p.headers.put(key, val);
+            
+            rest = line_data2.rest;
+        }
+        
+        var end_index: usize = 0; 
+        var end_trailer_index: usize = 0;
+        
+        if (p.headers.count() == 0 and hasPrefix(rest, pem_end[1..])) {
+            end_index = 0;
+            end_trailer_index = pem_end.len - 1;
+        } else {
+            end_index = mem.indexOf(u8, rest, pem_end) orelse 0;
+            end_trailer_index = end_index + pem_end.len;
+        }
+        
+        if (mem.indexOf(u8, rest, pem_end) == null) {
+            continue;
+        }
+        
+        var end_trailer = rest[end_trailer_index..];
+        const end_trailer_len = type_line.len + pem_end_of_line.len;
+        if (end_trailer.len < end_trailer_len) {
+            continue;
+        }
+
+        const rest_of_end_line = end_trailer[end_trailer_len..];
+        end_trailer = end_trailer[0..end_trailer_len];
+        if (!hasPrefix(end_trailer, type_line) or !hasSuffix(end_trailer, pem_end_of_line)) {
+            continue;
+        }
+        
+        const line_data2 = getLine(rest_of_end_line);
+        if (line_data2.line.len != 0) {
+            continue;
+        }
+        
+        const base64_data = try removeSpacesAndTabs(allocator, rest[0..end_index]);
+        const base64_decode_len = try std.base64.standard.Decoder.calcSizeForSlice(base64_data);
+
+        const decoded_data = try allocator.alloc(u8, base64_decode_len);
+        try std.base64.standard.Decoder.decode(decoded_data, base64_data);
+
+        p.bytes = decoded_data;
+
+        return p;
+    }
+}
+
+const nl = "\n";
+
+const pem_line_length = 64;
+
+fn writeHeader(writer: anytype, k: []const u8, v: []const u8) !void {
```

Comment:
You'd want to rename this to `fn appendHeader(list: *std.ArrayList(u8), ...)` because _a `writer`_ (at least as a convention) doesn't have a `appendSlice` method